File: examples/01-square-ice-shelf/square_ice_shelf.cpp
// In this example, we'll see how to use the basic functionality of icepack --
// creating a mesh, creating some scalar and vector fields, and solving the
// diagnostic equations for the velocity of an ice shelf.

// This file contains a bunch of import statements, definitions of physical
// constants, etc. -- it's not too relevant, so we've put it in a separate
// file.
#include "data.hpp"

int main()
{
  // First, we have to generate a mesh which discretizes the underlying
  // geometry. deal.II has a bunch of built-in functions for generating simple
  // meshes which we'll use here.
  dealii::Triangulation<2> mesh;
  const Point<2> p1(0.0, 0.0), p2(length, width);
  dealii::GridGenerator::hyper_rectangle(mesh, p1, p2);

  // Next, we need to mark the right-hand side of the geometry as belonging to
  // a different part of the boundary than the rest -- this is the ice front,
  // which has different boundary conditions.
  for (auto cell: mesh.active_cell_iterators())
    for (unsigned int face_number = 0;
         face_number < dealii::GeometryInfo<2>::faces_per_cell; ++face_number)
      if (cell->face(face_number)->center()(0) > length - 1.0)
        cell->face(face_number)->set_boundary_id(1);

  // Now refine the mesh a few times so that our solution will be decently
  // accurate.
  const unsigned int num_levels = 5;
  mesh.refine_global(num_levels);


  // Next we need to make some input data for the problem. The thickness is a
  // linear ramp which thins out towards the ice front.
  const Fn<2> thickness([&](const Point<2>& x)
                        {return h0 - delta_h/length * x[0];});

  // The velocity is defined in the data file -- it's the exact solution of the
  // diagnostic equations with the thickness above and a constant temperature
  // of -10C.
  const Velocity velocity;

  // But to keep things interesting, we'll perturb the temperature a bit in the
  // center of the domain.
  const Fn<2> temperature([&](const Point<2>& x)
                          {
                            const bool inside =
                              (x[0] > length/4 && x[0] < 3*length/4 &&
                               x[1] > width/4  && x[1] < 3*width/4);
                            return temp + inside * delta_temp;
                          });

  // Create a model object which knows how to solve the solve the equations of
  // motion for glacier flow.
  icepack::IceShelf ice_shelf(mesh, 1);

  // Interpolate the exact input fields to the finite element basis.
  const Field<2> theta = ice_shelf.interpolate(temperature);
  const Field<2> h = ice_shelf.interpolate(thickness);
  const VectorField<2> u0 = ice_shelf.interpolate(velocity);

  // The model object's diagnostic solve procedure takes in the ice thickness,
  // temperature and an initial guess for the velocity and computes a solution
  // of the shallow shelf equations. The velocity initial guess has to have the
  // right Dirichlet boundary conditions.
  const VectorField<2> u = ice_shelf.diagnostic_solve(h, theta, u0);

  // Write out the finite element fields to file so that we can do some post-
  // processing, visualization, etc.
  u.write("u.ucd", "u");
  u0.write("u0.ucd", "u");

  return 0;
}

File: examples/01-square-ice-shelf/square_ice_shelf.py

import numpy as np
import matplotlib.pyplot as plt
import icepack, icepack.plot

# Generate a mesh using `gmsh` and read it back in.
import make_mesh
length, width = 20.0e3, 20.0e3
make_mesh.main(length, width, "rectangle")


# The ice thickness decreases linearly from the inflow boundary to the calving
# front.
h0, dh = 500.0, 100.0
def thickness(x):
    return h0 - dh * x[0] / length


# Make the ice temperature jump by 5C in the middle of the domain.
temp = 254.15
def temperature(x):
    dtheta = 5.0
    inside = ((length/4 < x[0] < 3*length/4) and (width/4 < x[1] < 3*width/4))
    return temp + inside * dtheta


# The ice velocity is an exact solution for constant temperature and linear
# thickness. For the derivation, see the glacier mechanics book by Greve and
# Blatter.
def velocity(x):
    from icepack import rho_ice, rho_water, gravity, rate_factor
    u0 = 100.0
    rho = rho_ice * (1 - rho_ice / rho_water)
    A = (rho * gravity * h0 / 4)**3 * rate_factor(temp)
    q = 1 - (1 - (dh / h0) * (x[0] / length))**4
    return u0 + 0.25 * A * q * length * h0 / dh


# Read the mesh from a file, refine it a bit, and make a discretization.
mesh = icepack.read_msh("rectangle.msh")
mesh.refine_global(3)
discretization = icepack.make_discretization(mesh, 1)

# Interpolate all the exact data to the discretization we just made.
h = icepack.interpolate(discretization, thickness)
u0 = icepack.interpolate(discretization, velocity, lambda x: 0.0)
theta = icepack.interpolate(discretization, temperature)


# --- New stuff! ---

# The boundary segments of the mesh have different numeric IDs in order to help
# set where Dirichlet or Neumann boundary conditions apply. We can check what
# they are by plotting the mesh.
fig, ax = plt.subplots()
ax.set_aspect('equal')
icepack.plot.plot_mesh(ax, mesh)
plt.show(fig)

# From this figure we can see that boundary region 1 is where we want Dirichlet
# boundary conditions, and region 2 is where we want Neumann conditions.
dirichlet_boundary_ids = {1}

# Create a model object that stores these boundary conditions, along with all
# the silly configuration options for how to solve the nonlinear system.
ice_shelf = icepack.IceShelf(dirichlet_boundary_ids)

# Solve for the ice velocity!
u = ice_shelf.solve(h, theta, u0)

# Plot the result.
fig, ax = plt.subplots()
ax.set_aspect('equal')
icepack.plot.plot_field(ax, u)
plt.show(fig)

